### PR TITLE
Upload only linux_x86_64 wheels instead of manylinux2014

### LINKED ISSUE #776: Wheels are not ABI compatible with C++11
I move the discussion about the ABI compatibility of the new wheels (#774) from https://github.com/robotology/gym-ignition/pull/276.

Long story short. The standard `manylinux2014` [requires gcc4](https://www.python.org/dev/peps/pep-0599/#the-manylinux2014-policy) that does not support the new C++11 ABI for std::string. This means that the libraries included in the wheel included in the exported CMake project cannot be consumed by dowstream projects that use the new ABI (nowadays, mostly all the code).

The subtle part is that downstream compilation works just fine, but then the loader will not find any symbol that has the old `std::string` signature.

There are quite a lot of issues out there, an interesting one is https://github.com/pytorch/pytorch/issues/17492. Other interesting details are recapped in [How we build Apache Arrow's manylinux wheels](https://uwekorn.com/2019/09/15/how-we-build-apache-arrows-manylinux-wheels.html). A third one from tensorflow https://groups.google.com/a/tensorflow.org/g/build/c/Xo190xAPyW8?pli=1.

I'm not yet sure what is the best solution here. If supporting `manylinux2014` is too problematic, maybe uploading wheels not compliant with it would work just fine for the majority of the users. cc @traversaro

### PR DESCRIPTION
**Pull request by diegoferigo**
Closes #776

### CODE REVIEW

File: .github/workflows/python.yml [hunk 1]
```diff
@@ -88,6 +88,16 @@ jobs:
           CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014
           CIBW_BEFORE_BUILD: "yum install -y eigen3-devel assimp-devel libxml2-devel coin-or-Ipopt-devel swig3"
           CIBW_TEST_COMMAND: "python -c 'import idyntree.bindings'"
+          # Note: here we repair the wheel maintaining the linux_x86_64 platform
+          CIBW_REPAIR_WHEEL_COMMAND_LINUX: "auditwheel repair --plat linux_x86_64 -w {dest_dir} {wheel}"
+
+      # Rename the wheel to manylinux2014 (even if it is not) so that it can be uploaded to PyPI
```
Comment (traversaro) [inline .github/workflows/python.yml:94]: I hope the PyPI police will not come to look for us.

**Review by traversaro — APPROVED** [review]

File: .github/workflows/python.yml (hunk 1 above)
Comment (diegoferigo) [reply .github/workflows/python.yml:94]: I was not aware of this workaround, I think I read it the first time [here](https://thomastrapp.com/blog/building-a-pypi-package-for-a-modern-cpp-project/). We do the same also in other projects we maintain.

**Comment by diegoferigo** [comment]
Let me try the wheel before merging, I let you know if it is ok.

**Comment by traversaro** [comment]
Ok!

**Comment by diegoferigo** [comment]
In https://github.com/robotology/idyntree/issues/776#issuecomment-731524879 I wrote that is `auditwheel` that fixes the symbols to match the ABI. I was wrong. I think that `_GLIBCXX_USE_CXX11_ABI` is automatically set to iDynTree transitively from its dependencies found in the CentOS system.

In any case, using auditwheel to vendor the external dependencies would in any case result with a wheel with libraries with the old ABI. I think that there is no escape here.

**Let's revert to what we used to do until now: building on an ubuntu system and upload the wheel as it is, performing the renaming workaround.** We loose the vendoring of the dependencies, but this is the price we have to pay. I think we learnt a lot in these days and faced many of the challenges related to a PEP compliant PyPI packaging. There are proposals for what's coming after `manylinux2014` but nothing have been yet finalized. I guess we have to wait a new standard to strictly align with the packaging guidelines. Alternatively, conda is already available.

File: .github/workflows/python.yml (hunk 1 above)
Comment (diegoferigo) [reply .github/workflows/python.yml:93]: It is somehow documented in https://www.python.org/dev/peps/pep-0513/#compilation-of-compliant-wheels.

**Comment by traversaro** [comment]
Ok for me. Can you make a PR to refer to the old case? I think the automatic publishing on PyPI is still great, even if just in source form.

**Comment by diegoferigo** [comment]
You anticipated me, when you commented I was editing the comment above with the old code. Is it fine?

**Comment by diegoferigo** [comment]
Good to go as soon as CI turns green. I tested the wheels bu downloading the artifact and they work well :tada:

**Comment by traversaro** [comment]
Just to clarify for iDynTree on PyPI this is still  a change, right? 

Before we were posting on PyPI source archives, so anyone doing `pip install --pre idyntree` would download a source archive, and would compile it on his system, and the assumption was that it had the right dependencies to compile them (so on recent debian systems `libassimp-dev libxml2-dev libeigen3-dev coinor-libipopt-dev` . Now instead we compile and make available the Linux binaries compiled on Ubuntu 20.04, and so doing `pip install --pre idyntree` will download this linux binaries, and the assumption is that the user already has installed `libxml2 libassimp5 coinor-libipopt1v5` with the exact ABI that have on Ubuntu 20.04, so we assume that the user is on Ubuntu 20.04, right?

**Comment by traversaro** [comment]
To clarify, I am totally ok with this change, I just want to make sure that I got correctly change.

**Comment by diegoferigo** [comment]
> Just to clarify for iDynTree on PyPI this is still a change, right?

There is no real change, what you wrote here (with small modifications) is correct:

> Before we were posting **manually** on PyPI source archives, so anyone doing pip install --pre idyntree would download a source archive, and would compile it on his system, and the assumption was that it had the right dependencies to compile them (so on recent debian systems libassimp-dev libxml2-dev libeigen3-dev coinor-libipopt-dev . Now ~instead~ we **also** compile and make available the Linux binaries compiled on Ubuntu 20.04, and so doing pip install --pre idyntree will download these linux binaries, and the assumption is that the user already has installed libxml2 libassimp5 coinor-libipopt1v5 with the exact ABI that have on Ubuntu 20.04, so we assume that the user is on Ubuntu 20.04, right?

My point is that we upload both the `sdist` and the `linux_x86_64` wheel. By default, if both are available for a given platform, `pip` will prefer downloading the wheel. However, using `pip install idyntree --no-binary :all:`, the wheel can be ignored.

For the records, this is the expected ABI, that could be available also on other distributions, not specifically only for Ubuntu 20.04:

```
❯ ldd /rl/venv/lib/python3.8/site-packages/idyntree/_iDynTree.so 
        linux-vdso.so.1 (0x00007ffd8650b000)
        libassimp.so.5 => /lib/x86_64-linux-gnu/libassimp.so.5 (0x00007fb9307a3000)
        libxml2.so.2 => /lib/x86_64-linux-gnu/libxml2.so.2 (0x00007fb9305e9000)
        libipopt.so.1 => /lib/libipopt.so.1 (0x00007fb93037c000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fb93022d000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fb93004c000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fb93002f000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fb92fe3d000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fb92fe21000)
        libminizip.so.1 => /lib/x86_64-linux-gnu/libminizip.so.1 (0x00007fb92fc16000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fb92fc10000)
        libicuuc.so.66 => /lib/x86_64-linux-gnu/libicuuc.so.66 (0x00007fb92fa2a000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007fb92f9ff000)
        liblapack.so.3 => /lib/x86_64-linux-gnu/liblapack.so.3 (0x00007fb92f35b000)
        libblas.so.3 => /lib/x86_64-linux-gnu/libblas.so.3 (0x00007fb92f2ee000)
        libdmumps_seq-5.2.1.so => /lib/x86_64-linux-gnu/libdmumps_seq-5.2.1.so (0x00007fb92f0df000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fb93196d000)
        libicudata.so.66 => /lib/x86_64-linux-gnu/libicudata.so.66 (0x00007fb92d61e000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fb92d5f9000)
        libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007fb92d331000)
        libmumps_common_seq-5.2.1.so => /lib/x86_64-linux-gnu/libmumps_common_seq-5.2.1.so (0x00007fb92d2d9000)
        libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007fb92d28f000)
        libpord_seq-5.2.1.so => /lib/x86_64-linux-gnu/libpord_seq-5.2.1.so (0x00007fb92d276000)
        libesmumps-6.so => /lib/x86_64-linux-gnu/libesmumps-6.so (0x00007fb92d26e000)
        libscotch-6.so => /lib/x86_64-linux-gnu/libscotch-6.so (0x00007fb92d1d9000)
        libscotcherr-6.so => /lib/x86_64-linux-gnu/libscotcherr-6.so (0x00007fb92d1d4000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007fb92d1c1000)
```

**Comment by diegoferigo** [comment]
I updated my previous https://github.com/robotology/idyntree/pull/777#issuecomment-731586472 with new links and resources if anyone is interested in the upcoming `manylinux_${GLIBCMAJOR}_${GLIBCMINOR}_${ARCH}` platform.

**Comment by traversaro** [comment]
Ok thanks!

**Pull request merged**